fdbserver/StorageCache.actor.cpp

```diff
@@ -1911,25 +1910,19 @@ ACTOR Future<Void> pullAsyncData(StorageCacheData* data) {
 					           OTELSpanContextMessage::isNextIn(cloneReader)) {
 						OTELSpanContextMessage scm;
 						cloneReader >> scm;
-					} else if (cloneReader.protocolVersion().hasEncryptionAtRest() &&
-					           EncryptedMutationMessage::isNextIn(cloneReader) && !cipherKeys.present()) {
-						// Encrypted mutation found, but cipher keys haven't been fetch.
-						// Collect cipher details to fetch cipher keys in one batch.
-						EncryptedMutationMessage emm;
-						cloneReader >> emm;
-						cipherDetails.insert(emm.header.cipherTextDetails);
-						cipherDetails.insert(emm.header.cipherHeaderDetails);
-						collectingCipherKeys = true;
 					} else {
 						MutationRef msg;
-						if (cloneReader.protocolVersion().hasEncryptionAtRest() &&
-						    EncryptedMutationMessage::isNextIn(cloneReader)) {
-							assert(cipherKeys.present());
-							msg = EncryptedMutationMessage::decrypt(cloneReader, cloneReader.arena(), cipherKeys.get());
-						} else {
-							cloneReader >> msg;
+						cloneReader >> msg;
+						if (msg.isEncrypted()) {
+							if (!cipherKeys.present()) {
+								const BlobCipherEncryptHeader* header = msg.encryptionHeader();
+								cipherDetails.insert(header->cipherTextDetails);
+								cipherDetails.insert(header->cipherHeaderDetails);
+								collectingCipherKeys = true;
```

Comment:
nit: `cipherKeysHydrated` maybe? The current variable gives an impression that operation is a multi-step process, which it isn't is my understanding

Reply:
It is multi-step. Here we are iterating through a bunch of mutations from the TLog cursor, collect cipher key details, and make request to EKP to get all cipher keys, before start the loop again, where this time we have the cipher keys to decrypt the mutations.

Though per discussion with @sfc-gh-satherton offline, he convinces me to simplify this logic to do per-mutation fetching of cipher keys, which I will send PR for later.

Reply:
`Collecting` gives an impression that `collection` of information is a multi-step process :). Hence, my suggestion. 

However, if the plan is to fetch key one at a time, then for sure this logic can get simplified. Though it could mean more RPC calls to EKP, but, given SS local KeyCache, this should be an issue. 

Thanks

Reply:
The plan is for BlobCipherKeyCache to become a read-through cache that returns Futures.  Batching of RPCs behind that cache can happen internally if needed.

Caches returning Futures is the best way to provide a cache for asynchronous resources.  This is how both the SQLite and Redwood page caches work, as well as the ReadYourWrites cache, the open files map, and others.   The idea is for BlobCipherKeyCache to have an interface like
```
Future<Reference<BlobCipherKey>> getCipherKey(...);
```
which returns a Future that will be ready when the key is known. In the vast majority of cases, the returned future is already ready() when returned.  GetCipherKey() would be a function (not an actor) and would just check the cache for an existing Future, create/store a fetch actor if needed, and then return the cached future.  Such an actor could also internally batch requests by pausing to wait for more cache misses and combine them, but this would be internal to BlobCipherKeyCache and transparent to its users.  This can be a later optimization, as even with thousands of keys the _initial discovery_ of each key by each SS isn't that many RPCs to the EKP.  For example, SS's can serve almost 150k requests per second via its read endpoints.  I think batching is most important for the proactive _refresh_ of each cached key to prevent having to remove it from memory.

With the above API change, in the SS update loop there are no temporary maps and only 1 map lookup - in the actual key cache - per key needed.  This vastly simplifies the SS update loop and removes a bunch of helper functions as well as greatly reduces overhead.

Reply:
> The plan is for BlobCipherKeyCache to become a read-through cache that returns Futures. Batching of RPCs behind that cache can happen internally if needed.

BlobCipherKeyCache was designed as a library which can be used by all Encryption participant processes as a local cache for EncryptionKeys; I did think about making it a Actor but then due to unknown at that point especially around encrypting BlobGranules (part of the code is FDBClient which doesn't support actors), it felt better to design it as a "simple cache" and define an actor which then leverage cache to do the following:
1. Vectorized API to fetch encryption key(s)
2. First check the local processes cache (BlobCipherKeyCache) for presence and validity of the EK before reaching out to EKP (and/or KMS endpoint)

In current code, using GetCipherKeys actor supports both the usecases: 
a) allows client to obtain a Future<> which can then be _waited_ somepoint later in code execution, 
b) allows batching EKs lookup (performance optimization)

In theory we can merge GetCipherKeys into BlobCipherKeyCache even today and things would fine (given BlobGranule encryption usecase is better understood and implemented/merged as of today), OR, it can be used in its current form to achieve the same goals (GetCipherKeys actor abstracts out BlobCipherKeyCache from the caller).  
But, to me the change seems more of a code-cleanup and can be done later (but should be done to simplify the code). Please correct if am missing something. 

Thanks

Reply:
I'm not saying to change it for this PR, I was just explaining here what Yi and I had talked about offline.  Getting functionality is the short term goal for sure.  But the refactor should be done soon in a followup as SS CPU is a common bottleneck.  What I am proposing simplifies a lot of code and removes overhead in the common execution path so it's win-win.

> OR, it can be used in its current form to achieve the same goals (GetCipherKeys actor abstracts out BlobCipherKeyCache from the caller).
They should be merged, not separate, because the cache should hold futures at which point the wrapper is not needed.  With two classes, you have sort of a “read and then place into” cache instead of a “read through” cache which is less code and lower overhead.

The best way to cache an asynchronously fetched thing in Flow is to have a map of futures which implements a read-through cache.  Therefore, one key lookup == one cache container lookup and overlapping uncached requests do not cause duplicate RPCs.  Even in the cached case this is low overhead as Futures are reference counted and contain a T instance, so in the common case the cache container lookup gives you a reference to a future in the Ready state which already contains your T, which in this case a reference to an encryption key.

> Please correct if am missing something.
I think you were maybe missing the benefit of the read through cache pattern where you have a map of Futures by their lookup ID.  I hope I’ve cleared that up here, but if not let's talk offline.

If it helps, a good analog would be linux file descriptors and `read()`.  If multiple applications read the same location at the same time only one disk IO is dispatched, and it may even be combined with other outstanding reads of adjacent locations into a single disk IO (batching) but the caller can deal with a single-read function and not worry about that.  We could do the same with RPC batching for encryption keys but I doubt it is needed for cache misses, I think it's most important for the preemptive periodic refresh.

Reply:
@sfc-gh-satherton @sfc-gh-ahusain thank you both for the great discussion. I think my plan is to keep it as is for this PR, but will work on a PR later with what Steve suggests. I think the proposed change will make the code a lot readable at the least, and a big bonus if there's indeed measurable performance gain for less map lookup overhead.